### PR TITLE
add support for prank(sender, origin) and startPrank(sender, origin) cheatcodes

### PR DESCRIPTION
**Pull request by karmacoma-eth**

### CODE REVIEW

File: tests/regression/test/Prank.t.sol [hunk 1]
```diff
@@ -146,10 +203,9 @@ contract PrankTest is Test {
...
-    // function check_prank_ConstructorCreate2(address user, bytes32 salt) public {
-    //     vm.prank(user);
-    //     ConstructorRecorder recorder = new ConstructorRecorder{salt:salt}();
-    //     assert(recorder.caller() == user);
-    // }
+    function check_prank_ConstructorCreate2(address user, bytes32 salt) public {
+        vm.prank(user);
+        ConstructorRecorder recorder = new ConstructorRecorder{salt:salt}();
+        assert(recorder.caller() == user);
+    }
```
Comment (karmacoma-eth) [inline tests/regression/test/Prank.t.sol:210]: look at us, we do support CREATE2 now 🙌

**Comment by karmacoma-eth** [comment]
not sure what's going on with CI, some jobs flagging with:

```
 From https://github.com/Vectorized/multicaller
 * branch            b4a0dd037f1d770b2e9ae0b80bbd989707df43d0 -> FETCH_HEAD
fatal: Unable to find current revision in submodule path '../lib/multicaller/lib/forge-std'
fatal: Failed to recurse into submodule path '../lib/multicaller'
Your project has missing dependencies that could not be installed.
Error: 
failed to resolve file: "/home/runner/work/halmos/halmos/tests/regression/../lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Proxy.sol": No such file or directory (os error 2); check configured remappings
	--> /home/runner/work/halmos/halmos/tests/regression/test/Proxy.t.sol
	openzeppelin/proxy/ERC1967/ERC1967Proxy.sol
```

**Comment by karmacoma-eth** [comment]
> not sure what's going on with CI, some jobs flagging with:
> 
> ```
>  From https://github.com/Vectorized/multicaller
>  * branch            b4a0dd037f1d770b2e9ae0b80bbd989707df43d0 -> FETCH_HEAD
> fatal: Unable to find current revision in submodule path '../lib/multicaller/lib/forge-std'
> fatal: Failed to recurse into submodule path '../lib/multicaller'
> Your project has missing dependencies that could not be installed.
> Error: 
> failed to resolve file: "/home/runner/work/halmos/halmos/tests/regression/../lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Proxy.sol": No such file or directory (os error 2); check configured remappings
> 	--> /home/runner/work/halmos/halmos/tests/regression/test/Proxy.t.sol
> 	openzeppelin/proxy/ERC1967/ERC1967Proxy.sol
> ```

fixed itself with enough re-running 🤷‍♂️

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -53,4 +53,4 @@ jobs:
         run: python -m pip install -e .
 
       - name: Run pytest
-        run: pytest -n 4 -v -k "not long and not ffi" --ignore=tests/lib --halmos-options="-st ${{ matrix.parallel }} --storage-layout ${{ matrix.storage-layout }} --solver-timeout-assertion 0 ${{ inputs.halmos-options }}" ${{ inputs.pytest-options }}
+        run: pytest -n 1 -v -k "not long and not ffi" --ignore=tests/lib --halmos-options="--debug -st ${{ matrix.parallel }} --storage-layout ${{ matrix.storage-layout }} --solver-timeout-assertion 0 ${{ inputs.halmos-options }}" ${{ inputs.pytest-options }}
```
Comment (karmacoma-eth) [inline .github/workflows/test.yml:56]: `pytest -n 1` is what fixed CI, it seems that the issue was running multiple forge commands in parallel would result in race conditions trying to clone submodules or install solc

File: src/halmos/sevm.py [hunk 3]
```diff
@@ -1610,6 +1619,7 @@ def callback(new_ex: Exec, stack, step_id):
                 new_ex.jumpis = deepcopy(ex.jumpis)
                 new_ex.symbolic = ex.symbolic
                 new_ex.prank = deepcopy(ex.prank)
+                new_ex.origin = ex.origin
```
[the remark this answers is not in the document]
Comment (daejunpark) [reply src/halmos/sevm.py:1622]: probably we need more tests to cover these cases:
- checking tx.origin restored after contract creation
- checking tx.origin restored after test contract constructor when there is startPrank() but no closing stopPrank() in the constructor

File: src/halmos/sevm.py [hunk 4]
```diff
@@ -724,6 +726,13 @@ def current_opcode(self) -> UnionType[int, BitVecRef]:
     def current_instruction(self) -> Instruction:
         return self.pgm.decode_instruction(self.pc)
 
+    def resolve_prank(self, to: Address) -> Tuple[Address, Address]:
+        # this potentially "consumes" the active prank
+        prank_result = self.prank.lookup(to)
+        caller = self.this if prank_result.sender is None else prank_result.sender
+        origin = f_origin() if prank_result.origin is None else prank_result.origin
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/sevm.py:733]: I don't think so? `prank_result.origin is None` means that the origin must *not* be pranked, so we don't want the current pranked origin to be propagated

however this caused me to look deeper in the case of nested contexts and I do think we have an issue

[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/sevm.py:733]: here for instance, the fact that we create a new Prank() context for the sub_ex is wrong for nested contexts I think

https://github.com/a16z/halmos/blob/main/src/halmos/sevm.py#L1638-L1666

File: tests/regression/test/Prank.t.sol [hunk 5]
```diff
@@ -48,43 +64,117 @@ contract PrankSetUpTest is Test {
...
+    function check_prank_double(address user, address origin) public {
+        vm.prank(user, origin);
+
+        // the outer call is pranked
+        target.recordCaller();
+        assert(target.caller() == user);
+        assert(target.origin() == origin);
+
+        // but the inner call is not pranked
+        checkNotPranked(inner, address(target));
```
Comment (karmacoma-eth) [inline tests/regression/test/Prank.t.sol:112]: this also fails in foundry 🤨 is there an issue or is my mental model just wrong?

Comment (karmacoma-eth) [reply tests/regression/test/Prank.t.sol]: update: my mental model was wrong, and so the implementation was wrong too. I checked that all the tests in PrankTest now behave the same in both foundry and halmos ([df2c7a3](https://github.com/a16z/halmos/pull/336/commits/df2c7a3e77afa3712f10d1d5e6d89e5f802b8fdf))

File: src/halmos/sevm.py [hunk 6]
```diff
@@ -1847,8 +1858,8 @@ def create(
         if op == EVM.CREATE2:
             salt = ex.st.pop()
 
-        # lookup prank
-        caller = ex.prank.lookup(ex.this, con_addr(0))
+        # check if there is an active prank
+        pranked_caller, pranked_origin = ex.resolve_prank(address(ex.this))
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/sevm.py]: fixed in df2c7a3e77afa3712f10d1d5e6d89e5f802b8fdf

File: src/halmos/sevm.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/sevm.py]: fixed in df2c7a3e77afa3712f10d1d5e6d89e5f802b8fdf

File: src/halmos/sevm.py [hunk 7]
```diff
@@ -630,6 +630,7 @@ class Exec:  # an execution path
     jumpis: Dict[str, Dict[bool, int]]  # for loop detection
     symbolic: bool  # symbolic or concrete storage
     prank: Prank
+    origin: Address
```
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/sevm.py]: you're right, this is much simpler with origin being a field in CallContext that gets passed down to nested contexts

File: src/halmos/sevm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (karmacoma-eth) [reply src/halmos/sevm.py]: tests added in 1ea690c

**Review by daejunpark — APPROVED** [review]
I love the new method of maintaining the caller and origin using the call context and having them seamlessly interact with the prank context. Also appreciate the comprehensive tests! Thanks for the effort!

**Comment by daejunpark** [comment]
Note: issues #342 and #343 are orthogonal to this PR, so please feel free to merge this as is, and address them later.

File: src/halmos/cheatcodes.py [hunk 8]
```diff
@@ -75,58 +75,92 @@ def stringified_bytes_to_bytes(hexstring: str) -> ByteVec:
...
-    def startPrank(self, addr: Any) -> bool:
-        assert_address(addr)
-        if self.addr is not None:
-            return False
-        self.addr = addr
-        self.keep = True
-        return True
+    def startPrank(self, sender: Address, origin: Address | None = None) -> bool:
+        result = self.prank(sender, origin)
+        self.keep = result if result else self.keep
```
Comment (daejunpark) [inline src/halmos/cheatcodes.py:158]: This is accurate but a bit convoluted. For better code readability, I suggest having `prank()` take an additional argument (defaulting to `False` if not provided) that is assigned to `self.keep`. Then, have `startPrank()` call `prank()` with `keep=True`.

Comment (karmacoma-eth) [reply src/halmos/cheatcodes.py]: yeah you're right, in 17eba86

**Pull request merged**